packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,748 @@
+   */
+  dispose() {}
+}
+
+const SAMPLESIN = 1024;
+const SAMPLESOUT = 512;
+const inWaveData = new Float32Array(SAMPLESIN);
+const outSpectralData = new Float32Array(SAMPLESOUT);
+
+/**
+ * Feeds audio data to the FFT.
+ * @param analyser The AnalyserNode used to get the audio data.
+ * @param fft The FFTNullsoft instance from the PaintHandler.
+ */
+function processFFT(analyser: AnalyserNode, fft: FFT): void {
+  const dataArray = new Uint8Array(SAMPLESIN);
+
+  analyser.getByteTimeDomainData(dataArray);
+  for (let i = 0; i < dataArray.length; i++) {
+    inWaveData[i] = (dataArray[i] - 128) / 28;
+  }
+  fft.timeToFrequencyDomain(inWaveData, outSpectralData);
+}
+
+//? =============================== BAR PAINTER ===============================
+type PaintFrameFunction = () => void;
+type PaintBarFunction = (
+  ctx: CanvasRenderingContext2D,
+  x1: number,
+  x2: number,
+  barHeight: number,
+  peakHeight: number
+) => void;
+
+export class BarPaintHandler extends VisPaintHandler {
+  _analyser: AnalyserNode;
+  _fft: FFT;
+  _color: string = "rgb(255,255,255)";
+  _colorPeak: string = "rgb(255,255,255)";
+  // Off-screen canvas for pre-rendering a single bar gradient
+  _bar: HTMLCanvasElement = document.createElement("canvas");
+  _peak: HTMLCanvasElement = document.createElement("canvas");
+  _16h: HTMLCanvasElement = document.createElement("canvas"); // non-stretched
+  _bufferLength: number;
+  _dataArray: Uint8Array;
+  logged: boolean = false;
+  colorssmall: string[];
+  colorssmall2: string[];
+  paintBar: PaintBarFunction;
+  paintFrame: PaintFrameFunction;
+
+  constructor(vis: Vis) {
+    super(vis);
+    this._analyser = this._vis.analyser!;
+    this._fft = new FFT();
+    this._bufferLength = this._analyser.frequencyBinCount;
+    this._dataArray = new Uint8Array(this._bufferLength);
+
+    this.colorssmall = [
+      vis.colors[17],
+      vis.colors[14],
+      vis.colors[11],
+      vis.colors[8],
+      vis.colors[4],
+    ];
+    this.colorssmall2 = [
+      vis.colors[17],
+      vis.colors[16],
+      vis.colors[14],
+      vis.colors[13],
+      vis.colors[11],
+      vis.colors[10],
+      vis.colors[8],
+      vis.colors[7],
+      vis.colors[5],
+      vis.colors[4],
+    ];
+
+    this.logged;
+
+    this._16h.width = 1;
+    this._16h.height = 16;
+    this._16h.setAttribute("width", "75");
+    this._16h.setAttribute("height", "16");
+    if (this._vis.bandwidth === "wide") {
+      this.paintFrame = this.paintFrameWide.bind(this);
+    } else {
+      // thin
+      // does call paintFrameWide but there is a check inside
+      // that changes the bandwidth accordingly
+      this.paintFrame = this.paintFrameWide.bind(this);
+    }
+
+    if (this._vis.coloring === "fire") {
+      this.paintBar = this.paintBarFire.bind(this);
+    } else if (this._vis.coloring === "line") {
+      this.paintBar = this.paintBarLine.bind(this);
+    } else {
+      this.paintBar = this.paintBarNormal.bind(this);
+    }
```

Comment:
Personal style mostly, but I think `switch` is a bit cleaner for these.
```suggestion
    switch (this._vis.coloring) {
      case "fire":
        this.paintBar = this.paintBarFire.bind(this);
        break;
      case "line":
        this.paintBar = this.paintBarLine.bind(this);
        break;
      default:
        this.paintBar = this.paintBarNormal.bind(this);
        break;
   }
```

I think the same should apply to the cases below.